### PR TITLE
[node-agent] Add `Lease` controller for reporting health/liveness (similar to the `Lease` the `kubelet` writes)

### PR DESCRIPTION
**Pull request by majst01**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:

With this PR the node-agent writes a lease in the shoots kube-system namespace every 10s if everything works as expected, e.g. the node can be fetched.

**Which issue(s) this PR fixes**:
Part of #8023

**Special notes for your reviewer**:
/cc @rfranzke 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/nodeagent/controller/lease/node_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+package lease_test
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestNode(t *testing.T) {
```
Comment (shafeeqes) [inline pkg/nodeagent/controller/lease/node_suite_test.go]:
```suggestion
func TestLease(t *testing.T) {
```
and the filename

File: test/integration/nodeagent/lease/lease_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/gardener/gardener/pkg/logger"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+)
+
+func TestNode(t *testing.T) {
```
Comment (shafeeqes) [inline test/integration/nodeagent/lease/lease_suite_test.go]: Same here

File: pkg/nodeagent/controller/lease/node_suite_test.go (hunk 1 above)
Comment (majst01) [reply pkg/nodeagent/controller/lease/node_suite_test.go]: Fixed

File: test/integration/nodeagent/lease/lease_suite_test.go (hunk 2 above)
Comment (majst01) [reply test/integration/nodeagent/lease/lease_suite_test.go]: Fixed

File: pkg/nodeagent/controller/add.go [hunk 3]
```diff
@@ -46,5 +47,11 @@ func AddToManager(cancel context.CancelFunc, mgr manager.Manager, cfg *config.No
 		return fmt.Errorf("failed adding token controller: %w", err)
 	}
 
+	if err := (&lease.Reconciler{
```
Comment (rfranzke) [inline pkg/nodeagent/controller/add.go]: Please add some short documentation for this controller to https://github.com/gardener/gardener/blob/master/docs/concepts/node-agent.md#controllers (similar to https://github.com/gardener/gardener/blob/master/docs/concepts/gardenlet.md#lease-reconciler)

File: pkg/nodeagent/controller/add.go [hunk 4]
```diff
@@ -46,5 +47,11 @@ func AddToManager(cancel context.CancelFunc, mgr manager.Manager, cfg *config.No
...
+		HostName: nodeName,
```
Comment (rfranzke) [inline pkg/nodeagent/controller/add.go]: I don't think the hostname of the VM is equal to the node name in all cases. In my branch, I have restricted the `node-agent`'s manager cache to only `Node`s which are labeled with `kubernetes.io/hostname=<hostname-of-node-the-gna-runs-on>`, see https://github.com/rfranzke/gardener/blob/gna/gardenlet/cmd/gardener-node-agent/app/app.go#L184.

Hence, you don't need this `HostName` attribute on the reconciler at all, and you also don't need a `NodePredicate` for filtering for the "correct" node (since the correct node is only one in the manager's cache).

```suggestion
```

File: pkg/nodeagent/controller/lease/add.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	if r.RenewIntervalSeconds == 0 {
+		r.RenewIntervalSeconds = 60
+	}
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.Namespace == "" {
+		r.Namespace = metav1.NamespaceSystem
+	}
+	r.leaseName = "gardener-node-agent-" + r.HostName
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/add.go]:
```suggestion
```

File: pkg/nodeagent/controller/lease/add.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+// ControllerName is the name of the controller.
+const ControllerName = "lease"
+
+// AddToManager adds the lease controller with the default Options to the manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+	if r.RenewIntervalSeconds == 0 {
+		r.RenewIntervalSeconds = 60
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/add.go]: Where are these `60s` coming from? Is this the same interval the `kubelet` uses?

File: pkg/nodeagent/controller/lease/add.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	r.leaseName = "gardener-node-agent-" + r.HostName
+
+	node := &metav1.PartialObjectMetadata{}
+	node.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Node"))
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(node, builder.WithPredicates(r.NodePredicate())).
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/add.go]:
```suggestion
		For(node).
```

File: pkg/nodeagent/controller/lease/add.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+func (r *Reconciler) NodePredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Object.GetName() == r.HostName
+		},
+		UpdateFunc:  func(_ event.UpdateEvent) bool { return false },
+		DeleteFunc:  func(_ event.DeleteEvent) bool { return false },
+		GenericFunc: func(_ event.GenericEvent) bool { return false },
+	}
+}
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/add.go]:
```suggestion
```

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Reconciler creates a lease in the kube-system namespace of the shoot.
+type Reconciler struct {
+	Client               client.Client
+	RenewIntervalSeconds int32
+	leaseName            string
+	HostName             string
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]:
```suggestion
```

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Reconciler creates a lease in the kube-system namespace of the shoot.
+type Reconciler struct {
+	Client               client.Client
+	RenewIntervalSeconds int32
+	leaseName            string
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]:
```suggestion
```

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+	node := &metav1.PartialObjectMetadata{}
+	node.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Node"))
+	if err := r.Client.Get(ctx, request.NamespacedName, node); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	lease := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.leaseName,
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]:
```suggestion
			Name:      "gardener-node-agent-" + node.GetName(),
```

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+			Namespace: r.Namespace,
+		},
+	}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(lease), lease); err != nil {
+		if apierrors.IsNotFound(err) {
+			lease.Spec = coordinationv1.LeaseSpec{
+				HolderIdentity:       &r.leaseName,
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]:
```suggestion
				HolderIdentity:       &lease.Name,
```

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+				LeaseDurationSeconds: &r.RenewIntervalSeconds,
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]: When `r.RenewIntervalSeconds` is used for `LeaseDurationSeconds`, shouldn't we then maybe rename it to make its purpose more clear?

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+	lease.Spec = coordinationv1.LeaseSpec{
+		HolderIdentity:       &r.leaseName,
+		LeaseDurationSeconds: &r.RenewIntervalSeconds,
+		RenewTime:            &metav1.MicroTime{Time: r.Clock.Now().UTC()},
+	}
+
+	log.V(1).Info("Renewing heartbeat Lease", "lease", client.ObjectKeyFromObject(lease))
+	return reconcile.Result{
+		RequeueAfter: time.Duration(r.RenewIntervalSeconds) * time.Second / 3,
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]: Why `/3`? 👀

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+			return reconcile.Result{RequeueAfter: time.Duration(r.RenewIntervalSeconds) * time.Second}, r.Client.Create(ctx, lease)
+		}
+		return reconcile.Result{}, err
+	}
+
+	lease.Spec = coordinationv1.LeaseSpec{
+		HolderIdentity:       &r.leaseName,
+		LeaseDurationSeconds: &r.RenewIntervalSeconds,
+		RenewTime:            &metav1.MicroTime{Time: r.Clock.Now().UTC()},
+	}
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]: What about storing this desired spec in a variable before L59 and then just setting it here and in L61?

File: test/integration/nodeagent/lease/lease_test.go [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+
+	leasecontroller "github.com/gardener/gardener/pkg/nodeagent/controller/lease"
+	gardenerutils "github.com/gardener/gardener/pkg/utils"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+)
+
+var _ = Describe("Reconcile", func() {
+
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
```

File: test/integration/nodeagent/lease/lease_test.go [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+	Describe("Lease controller tests", func() {
+		var (
+			node      *corev1.Node
+			nodeName  string
+			fakeclock *testing.FakeClock
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
			fakeClock *testing.FakeClock
```

File: test/integration/nodeagent/lease/lease_test.go [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+			DeferCleanup(func() {
+				By("Delete Node")
+				Expect(client.IgnoreNotFound(testClient.Delete(ctx, node))).To(Succeed())
+				By("Delete Lease")
+				lease := &coordinationv1.Lease{ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}}
+				Expect(client.IgnoreNotFound(testClient.Delete(ctx, lease))).To(Succeed())
+			})
+		})
+
+		It("Reconcile should create lease", func() {
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
		It("should create Lease", func() {
```

File: test/integration/nodeagent/lease/lease_test.go [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+				Expect(testClient.Get(ctx, types.NamespacedName{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}, lease)).To(Succeed())
+				return lease.Spec.RenewTime.Equal(oldRenewTime)
+			}).Should(BeTrue())
+			validateOwnerReference(lease, node)
+		})
+	})
+})
+
+func validateOwnerReference(lease *coordinationv1.Lease, node *corev1.Node) {
+	Expect(lease.OwnerReferences).To(ConsistOf(metav1.OwnerReference{
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
	ExpectWithOffset(1, lease.OwnerReferences).To(ConsistOf(metav1.OwnerReference{
```

ref https://github.com/gardener/gardener/blob/master/docs/development/testing.md#writing-tests-relevant-for-all-kinds

File: test/integration/nodeagent/lease/lease_test.go [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+		It("Reconcile should create lease", func() {
+			lease := &coordinationv1.Lease{}
+			Eventually(func() error {
+				return testClient.Get(ctx, types.NamespacedName{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}, lease)
+			}).Should(Succeed())
+			validateOwnerReference(lease, node)
+		})
+
+		It("Reconcile should update lease time", func() {
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
		It("should update the Lease time", func() {
```

File: test/integration/nodeagent/lease/lease_test.go [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+			lease := &coordinationv1.Lease{}
+			Eventually(func() error {
+				return testClient.Get(ctx, types.NamespacedName{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}, lease)
+			}).Should(Succeed())
+			validateOwnerReference(lease, node)
+			oldRenewTime := lease.Spec.RenewTime
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go:71]:
```suggestion

			oldRenewTime := lease.Spec.RenewTime
```

File: test/integration/nodeagent/lease/lease_test.go [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+			fakeclock.SetTime(fakeclock.Now().Add(1 * time.Minute))
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]: Use `Step` instead of `SetTime`?

Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]: Why `1m`? Shouldn't we use something closer to the configured `RenewIntervalSeconds` to make the assertion below more accurate?

File: test/integration/nodeagent/lease/lease_test.go [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+
+			Eventually(func() bool {
+				Expect(testClient.Get(ctx, types.NamespacedName{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}, lease)).To(Succeed())
+				return lease.Spec.RenewTime.After(oldRenewTime.Time)
+			}).Should(BeTrue())
+			validateOwnerReference(lease, node)
+		})
+
+		It("Reconcile should not update lease time if no node is present", func() {
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
		It("should not update the Lease time if no node is present", func() {
```

File: test/integration/nodeagent/lease/lease_test.go [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+			lease := &coordinationv1.Lease{}
+			Eventually(func() error {
+				return testClient.Get(ctx, types.NamespacedName{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}, lease)
+			}).Should(Succeed())
+			validateOwnerReference(lease, node)
+			oldRenewTime := lease.Spec.RenewTime
+			fakeclock.SetTime(fakeclock.Now().Add(1 * time.Minute))
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]: same here

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks for the PR!
Had a joint review with @ScheererJ

File: pkg/nodeagent/controller/lease/add.go (hunk 8 above)
Comment (oliver-goetz) [reply pkg/nodeagent/controller/lease/add.go]: This predicate should prevent the reconciler to update the lease on node updates. Create event should start the reconciler, then it should be reconciled after `RenewIntervalSeconds` only.
Without the predicate, frequent updates of the node lead to frequent updates of the lease.

File: pkg/nodeagent/controller/lease/reconciler.go (hunk 14 above)
Comment (oliver-goetz) [reply pkg/nodeagent/controller/lease/reconciler.go]: The renewal interval should be shorter then the lease duration.
Instead of maintaining two different intervals we made this implicit contract.

The variable name should be `leaseDuration` though.

Comment (oliver-goetz) [reply pkg/nodeagent/controller/lease/reconciler.go]: Btw. kubelet is doing something similar ([ref](https://github.com/kubernetes/kubernetes/blob/36930f4cb605450ccbe3126caad365aa5f51267c/pkg/kubelet/kubelet.go#L891)). There it is `leaseDuration * 0.25`.

**Comment by majst01** [comment]
Addressed most of the comments except the removal of the NodePredicate.

File: pkg/nodeagent/controller/lease/add.go (hunk 6 above)
Comment (oliver-goetz) [reply pkg/nodeagent/controller/lease/add.go]: It was a best guess, that the lease is updated frequent enough to see in time that something is going wrong with the node agent, but not creating to many requests to the API server.
With 60 seconds and the `/3` calculation I explained in the other comment, node-agent would update the lease every 20 seconds.

kubelet has a default lease duration of 40 seconds ([ref](https://github.com/kubernetes/kubernetes/blob/36930f4cb605450ccbe3126caad365aa5f51267c/pkg/kubelet/apis/config/v1beta1/defaults.go#L132-L134)) and typically renews it every 10 seconds (`0.25* lease duration).

File: pkg/nodeagent/controller/add.go (hunk 3 above)
Comment (rfranzke) [reply pkg/nodeagent/controller/add.go]: Still missing

File: pkg/nodeagent/controller/add.go [hunk 25]
```diff
@@ -46,5 +47,11 @@ func AddToManager(cancel context.CancelFunc, mgr manager.Manager, cfg *config.No
 		return fmt.Errorf("failed adding token controller: %w", err)
 	}
 
+	if err := (&lease.Reconciler{
+		NodeName:      nodeName,
```
Comment (rfranzke) [inline pkg/nodeagent/controller/add.go]:
```suggestion
```

File: pkg/nodeagent/controller/lease/add.go (hunk 6 above)
Comment (rfranzke) [reply pkg/nodeagent/controller/lease/add.go]: Why don't we use the same settings like kubelet then? Or can you present a good reason for using `60s` instead of `40s`?

File: pkg/nodeagent/controller/lease/add.go (hunk 8 above)
Comment (rfranzke) [reply pkg/nodeagent/controller/lease/add.go]: This can be easily achieved like this: https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/pkg/controllermanager/controller/event/add.go#L43

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Reconciler creates a lease in the kube-system namespace of the shoot.
+type Reconciler struct {
+	Client               client.Client
+	RenewIntervalSeconds int32
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]:
```suggestion
	LeaseDurationSeconds int32
```

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+	NodeName             string
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]:
```suggestion
```

File: pkg/nodeagent/controller/lease/reconciler.go (hunk 14 above)
Comment (rfranzke) [reply pkg/nodeagent/controller/lease/reconciler.go]: Then let's also do this. Why are we doing `/3` and not `/4`?

File: test/integration/nodeagent/lease/lease_test.go [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+
+			go func() {
+				defer GinkgoRecover()
+				Expect(mgr.Start(mgrContext)).To(Succeed())
+			}()
+
+			DeferCleanup(func() {
+				By("Stop manager")
+				mgrCancel()
+			})
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]: We could move this part to `lease_suite_test.go` since it's enough to do this once per test suite.

File: test/integration/nodeagent/lease/lease_test.go [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+				},
+			}
+
+			By("Create Node")
+			Expect(testClient.Create(ctx, node)).To(Succeed())
+			DeferCleanup(func() {
+				By("Delete Node")
+				Expect(client.IgnoreNotFound(testClient.Delete(ctx, node))).To(Succeed())
+				By("Delete Lease")
+				lease := &coordinationv1.Lease{ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}}
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]: For consistency reasons, I would move this line below L87

File: test/integration/nodeagent/lease/lease_test.go [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+				Expect(client.IgnoreNotFound(testClient.Delete(ctx, lease))).To(Succeed())
+			})
+		})
+
+		It("should create Lease", func() {
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]:
```suggestion
		It("should create the Lease", func() {
```

File: pkg/nodeagent/controller/lease/add.go (hunk 8 above)
Comment (oliver-goetz) [reply pkg/nodeagent/controller/lease/add.go]: In my opinion it is to fragile to rely on the cache only. `node-agent` controllers (lease, node, healthcheck) are designed to reconcile one node only and will start to do weird stuff, in case they would reconcile more than one node.  
`lease` controller could create its leases with OwnerReferences to the wrong node, so that they are garbage collected at the wrong time.
`node` controller could react on annotations of the wrong node, so a restart annotation could cause a restart of a systemd unit at a random node.
`healthcheck` controller could even add its own internal IP to a different node status.

Without predicates for the node names (better labels) you don't see that, when you look at the controllers. 
You and I know that now, but it could be forgotten e.g. when the `controller-runtime` cache implementation is changed again in a year or two. No test would discover that, because the managers of integration tests have a dedicated cache configuration.
There won't be a measurable increase of compute resource. Of course, it adds a couple of code lines, but those will also document, that those are single node controllers.

Comment (rfranzke) [reply pkg/nodeagent/controller/lease/add.go]: I cannot follow this argument. If we have the tools at hand (like configuring the cache), not using them because "they could change in a year or two" doesn't make much sense to me (unless they are clearly stated as "fragile"/subject to change by the authors - which isn't the case for the manager caches).
If you want, feel free to write a small integration test that ensures that the cache works as expected, but TBH, I don't think this is necessary/over-engineered.
What makes you think the controller-runtime would change their cache implementation in such a breaking way that we cannot adapt to it?

Comment (rfranzke) [reply pkg/nodeagent/controller/lease/add.go]: Ah, and one point: We don't know the node name upfront for new VMs. We only know the hostname, but when kubelet creates the `Node`, its name could differ. Hence, when really wanted to introduce such predicates, we would need to wait for the `OperatingSystemConfig`s first successful reconciliation (which installs `kubelet` for the first time) and only then register the other controllers (with predicate of correct node name) to the manager. Not a big deal, but definitely complexity and a special code path. I would really like to avoid all this.

File: pkg/nodeagent/controller/lease/add.go (hunk 6 above)
Comment (majst01) [reply pkg/nodeagent/controller/lease/add.go]: I will change it to the same values as the kubelet

File: pkg/nodeagent/controller/lease/add.go (hunk 8 above)
Comment (oliver-goetz) [reply pkg/nodeagent/controller/lease/add.go]: We could use the [nodeName](https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/cmd/gardener-node-agent/app/app.go#L150) `node-agent` determines during startup for the predicates too.

In my opinion the cache configuration is meant to configure the cache in an efficient way (e.g. to consume as little memory, CPU as possible). We should definitely do that in `node-agent`.
However, the controller itself should be responsible to ensure that its business logic works properly. This also means that it restricts the objects it is reconciling to those it is responsible for. There will be more code, but it will be more resilient towards undesired side-effects.

I don't want to test `controller-runtime`. This is just an example for something which might go wrong in the future. It could also happen when refactoring `node-agent`. I'm pretty sure that we could adapt our implementation in those cases. The question is "Do we always remember that we have to adapt our implementation?". I cannot name a specific event when (or if) this happens, it is a risk though.

Comment (rfranzke) [reply pkg/nodeagent/controller/lease/add.go]: > We could use the [nodeName](https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/cmd/gardener-node-agent/app/app.go#L150) `node-agent` determines during startup for the predicates too.

We can't - this has changed in the branch where gardenlet integrates GNA into the shoot flow, see [here](https://github.com/rfranzke/gardener/commit/9c6dc8247c2b988b90cb32ee3fe52bde4d846db7#diff-55fb15b9da09f83f4983b2003c189b5e2ad4d71afb010bbf61f97dfe1a0b8fcf) the changes in GNA's `app.go` file. When it starts, the node name is not known yet.

> In my opinion the cache configuration is meant to configure the cache in an efficient way (e.g. to consume as little memory, CPU as possible). We should definitely do that in `node-agent`. However, the controller itself should be responsible to ensure that its business logic works properly. This also means that it restricts the objects it is reconciling to those it is responsible for. There will be more code, but it will be more resilient towards undesired side-effects.

The cache options can be used to restrict which objects the component can see. If there is only one object (or a set of objects) relevant to the component which can be perfectly filtered already here, why should we extend each and every controller to double-check again for the "right/expected" set? I really cannot comprehend this argumentation.

We already follow this approach in other components like `gardenlet`, e.g. when it filters for resources only relevant for the seed it is responsible for. [Here](https://github.com/gardener/gardener/blob/524084c5fb31a9b4e4642273f10e96c3e9a2badf/cmd/gardenlet/app/app.go#L262-L264), for example, `ControllerInstallation`s are already filtered based on the `.spec.seedRef.name`, so that we don't need any [additional predicates](https://github.com/gardener/gardener/blob/524084c5fb31a9b4e4642273f10e96c3e9a2badf/pkg/gardenlet/controller/controllerinstallation/controllerinstallation/add.go#L61-L64) in the respective controller(s).

> I don't want to test `controller-runtime`. This is just an example for something which might go wrong in the future. It could also happen when refactoring `node-agent`. I'm pretty sure that we could adapt our implementation in those cases. The question is "Do we always remember that we have to adapt our implementation?". I cannot name a specific event when (or if) this happens, it is a risk though.

Mistakes can and will happen either way. We could also "accidentally" drop the predicate and corresponding tests during a refactoring. With the same probability as newer versions of `controller-runtime` could introduce bugs in the cache options, they could also introduce bugs in the predicate handling. In the end, we rely on this third-party component and must trust that it works as expected. If we don't have this trust, we cannot safely make use of its features, and the logical consequence is then to rewrite the critical parts on our own. I guess nobody wants that.

In summary, the main argument for me is: If we can already filter the relevant objects globally via cache options, there is neither need nor benefit for doing it again and again in each and every single controller. In GNA, we already have two (and soon three) controllers reconciling `Node` objects, and I don't see value in duplicating such check.
Risk that people break stuff during code changes exists always. If you want to be extra sure, you can add a unit test for the `app.go` file that asserts that the manager is started with the expected cache options, but for me this is already far beyond the "tangible". Hence, I vote for keeping it simple and straight-forward.

Comment (rfranzke) [reply pkg/nodeagent/controller/lease/add.go]: > > We could use the [nodeName](https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/cmd/gardener-node-agent/app/app.go#L150) `node-agent` determines during startup for the predicates too.
> 
> We can't - this has changed in the branch where gardenlet integrates GNA into the shoot flow, see [here](https://github.com/rfranzke/gardener/commit/9c6dc8247c2b988b90cb32ee3fe52bde4d846db7#diff-55fb15b9da09f83f4983b2003c189b5e2ad4d71afb010bbf61f97dfe1a0b8fcf) the changes in GNA's `app.go` file. When it starts, the node name is not known yet.

Here is the PR that adapts the handling regarding the node determination: https://github.com/gardener/gardener/pull/8826

Btw, I saw that we already rely on the cache options for the existing `Node` controller. We don't any predicate regarding the "correct" node there: https://github.com/gardener/gardener/blob/eba989bc166232707bb1309c7df61cc6fa1ee8ca/pkg/nodeagent/controller/node/add.go#L47-L69

I hope you are okay with continuing with this approach as motivated/justified in the previous comment.

File: docs/concepts/node-agent.md [hunk 31]
```diff
@@ -32,6 +32,14 @@ It also executes tasks that need to be executed before any other components are
 
 This section describes the controllers in more details.
 
+### [`Lease` Controller](../../pkg/nodeagent/controller/lease)
+
+This controller creates a lease for `gardener-node-agent` in `kube-system` namespace of the shoot cluster.
```
Comment (rfranzke) [inline docs/concepts/node-agent.md]:
```suggestion
This controller creates a `Lease` for `gardener-node-agent` in `kube-system` namespace of the shoot cluster.
```

File: docs/concepts/node-agent.md [hunk 32]
```diff
@@ -32,6 +32,14 @@ It also executes tasks that need to be executed before any other components are
...
+Each instance of `gardener-node-agent` creates its own lease when its corresponding `Node` was created.
```
Comment (rfranzke) [inline docs/concepts/node-agent.md]:
```suggestion
Each instance of `gardener-node-agent` creates its own `Lease` when its corresponding `Node` was created.
```

File: docs/concepts/node-agent.md [hunk 33]
```diff
@@ -32,6 +32,14 @@ It also executes tasks that need to be executed before any other components are
...
+It renews the `Lease` resource every 10 seconds. This indicates a heartbeat to the external world.
+
+At the moment, the `Lease` resources are not reflected in any health check function of the shoot cluster yet.
```
Comment (rfranzke) [inline docs/concepts/node-agent.md]: I would remove this sentence - we might easily forget about it, and we plan to work on this anyways soon. Also, this document should rather focus on what the controller does and why, but not how others use it :)

File: pkg/nodeagent/controller/lease/reconciler.go [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+			log.V(1).Info("Creating heartbeat Lease", "lease", client.ObjectKeyFromObject(lease))
+			return reconcile.Result{RequeueAfter: requeueAfter}, r.Client.Create(ctx, lease)
+		}
+		return reconcile.Result{}, err
+	}
+
+	lease.Spec = leaseSpec
+
+	log.V(1).Info("Renewing heartbeat Lease", "lease", client.ObjectKeyFromObject(lease))
+	return reconcile.Result{RequeueAfter: requeueAfter}, r.Client.Update(ctx, lease)
```
Comment (rfranzke) [inline pkg/nodeagent/controller/lease/reconciler.go]: Can we use `controllerutil.CreateOrUpdate` here?

File: test/integration/nodeagent/lease/lease_test.go [hunk 35]
```diff
@@ -0,0 +1,116 @@
...
+		It("should update the Lease time", func() {
+			lease := &coordinationv1.Lease{}
+			Eventually(func() error {
+				return testClient.Get(ctx, types.NamespacedName{Namespace: testNamespace.Name, Name: "gardener-node-agent-" + nodeName}, lease)
+			}).Should(Succeed())
+			validateOwnerReference(lease, node)
+
+			oldRenewTime := lease.Spec.RenewTime
+			// wait a nit more than LeaseDurationSeconds
+			fakeClock.Step(5 * time.Second)
```
Comment (rfranzke) [inline test/integration/nodeagent/lease/lease_test.go]: Can you store the `LeaseDurationSeconds` in a variable and use it here and in other occurrences? This way, when somebody changes the value, they wouldn't need to adapt all these places here as well.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Nice, minor nits only left

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
